### ISSUE TITLE
 @As and @CaseDescription annotation use different placeholder enumerations

### ISSUE THREAD

**Issue by cirquit** [TNG/JGiven#282]
If I want to adress the first argument in the `@As ( "First argument: $1, Second Argument: $2" )`, it starts with the index 1. In **@CaseDescription** it starts with index 0.

I would propose to set the **@As** index to start from 0, because usually programmer will write these tests.

Need to add at least two examples for simple placeholder and enumerated ones to docs and jgiven-examples.

`@As ( "First argument: $, second argument $")`
`@As ( "Second argument: $1, first argument $0")`

`@CaseDescription ( "First argument: $, second argument $")`
`@CaseDescription ( "Second argument: $1, first argument $0")`

CaseDescription is also not documented yet.

Minimal example for **@CaseDescription** is found [here](https://github.com/TNG/JGiven/blob/master/jgiven-junit/src/test/java/com/tngtech/jgiven/junit/JUnitParamsTest.java).

Minimal example for **@As** Description:
``` java

import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import org.junit.Test;

import com.tngtech.jgiven.Stage;
import com.tngtech.jgiven.annotation.As;
import com.tngtech.jgiven.annotation.Description;
import com.tngtech.jgiven.annotation.IntroWord;
import com.tngtech.jgiven.junit.SimpleScenarioTest;
import org.junit.runner.RunWith;

RunWith( JUnitParamsRunner.class )
@Description( "Demonstrates the usage of the @As annotation" )
public class AsAnnotationExampleTest extends SimpleScenarioTest<AsAnnotationExampleTest.AsAnnotationStage> {

    @Test
    @As( "Scenario that shows the usage of @At with argument enumeration")
    @Parameters({
            "false, 0"
    })
    public void steps_can_use_at_annotation_to_reference_arguments_by_enumeration(boolean bool, int i) {
        given().some_boolean_and_int_value( bool, i );
    }

    public static class AsAnnotationStage extends Stage<AsAnnotationStage> {

        @As( "Referencing first argument : $1, second argument : $2 ")
        public AsAnnotationStage some_boolean_and_int_value( boolean bool, int i) {
            return this;
        }
    }
}
```

**Comment by Airblader** [comment]
This is unfortunate, but I'd prefer starting at 1. I expect a programmer to know both, but not somebody else. It also feels more natural.

**Comment by janschaefer** [comment]
I have a crazy idea: what if we just support both? In nearly all cases it is unambiguous what the user actually meant. Examples:

* User uses $0 for any parameter -> $0 must be the first one
* Users uses $1 and there is only one parameter -> $1 must be the first one
* Users uses $1 and $2 and there are two parameters -> $1 must be the first one

The only case where it is ambiguous is when there are more parameters than the user actually refers to and the user does not use $0, but $1. In that case $1 could be the first or the second parameter.

**Comment by janschaefer** [comment]
Alternatively, we could introduce a configuration option for that.

**Comment by Airblader** [comment]
That sounds like an unnecessarily complex mental model for my taste. As a user I expect the same variable to mean the same thing across tests. I shouldn't have to worry about changing behavior because I copy/paste a description and then remove $0, IMHO. 

I don't think this is worth configuration either. We should choose one of the two options and stick with it. :-)

**Comment by janschaefer** [comment]
Well I agree. In that case, we should stick with `$1` to be the first one. As the `@As` annotation is much more used than the `@CaseDescription` annotation.

**Comment by janschaefer** [comment]
We could replace the `@CaseDescription` annotation with a new one and deprecate it, to have a consistent behavior in the long run.

**Comment by Airblader** [comment]
Yeah, that sounds reasonable to me. We could also print a warning for using the deprecated annotation.

**Comment by janschaefer** [comment]
Well, the IDE will print the warning already :-)

**Comment by janschaefer** [comment]
A good argument for starting with 1 is that it is consistent with regular expression back references that also start at 1 :-)

**Comment by Airblader** [comment]
Forgot about the IDE. Too little Java lately :-( You're right of course, that should suffice.

**Comment by cirquit** [comment]
Ok, then I'm gonna use the index starting from 1 in the [issue 236](../issues/236) for ExtendedDescription and try to tackle this after that one.

**Comment by janschaefer** [comment]
I scheduled this issue for Milestone v0.15.0

**Comment by cirquit** [comment]
How should the new `CaseDescription` annotation with the enumeration starting from 1 be called?

**Comment by nikowitt** [comment]
I suggest something like useOrdinalPositionCount.

**Comment by cirquit** [comment]
Did you mean that for the possible configuration option? This would mean that the `CaseDescription` should take this as another argument and for everybody that uses it in the current form there would be a default?

I thought that creating a new annotation like OrdinalCaseDescription would be the proper thing to do, but then the `As` annotation does not have this prefix and is still ordinal...

**Comment by janschaefer** [comment]
Fixed by #301
